### DISCUSSION TITLE
`PHP 8.3` support.

### DISCUSSION THREAD

**Discussion by terabytesoftw** [yiisoft/yii2#19957] [General]
Hi, we should consider upgrading the `PHP` version in the master branch. Running `PHPUnit 4` to accommodate new `PHP` versions has become quite cumbersome due to the need to patch `PHPUnit`, which sometimes leads to unexpected errors. By raising the minimum required `PHP` version to at least `8.1`, in my opinion, we can update our packages without breaking backward compatibility. In my tests with branch `2.2`, i found that `PHP 8.3` is entirely compatible. Those who need to use versions older than `8.1` can continue using `2.0.49`, as suggested by @rob006 https://github.com/yiisoft/yii2/discussions/19831#discussioncomment-6988915.

This would, in some way, allow us to maintain the code more easily and address bugs more effectively.

Test php 8.3:  https://github.com/yiisoft/yii2/pull/19953

@yiisoft/yii2, @yiisoft/reviewers

**Comment by bizley** [comment]
I like the idea but it put some technical burden on us with versions and fixes for 2.0.49, basically it would have to be 2.0.49.1 and so on. I know we are doing it right now and we are not semver compatible but still, just to remember.

**Comment by rob006** [comment]
@bizley I proposed this **instead** of releasing 2.2. In that case it should be less work than currently planned for 2.2.

With PHP 7.3 as a new minimum we could use phpunit 9 and cover ~90% of users. In that case 2.0.49.x could receive only security fixes, and master branch could be developed more easily than now.

**Comment by terabytesoftw** [comment]
I think it is useful to keep the `2.2` branch, i will continue removing the deprecations, and i will keep the branch, once the open PR is merged, i will continue working on it.

**Comment by schmunk42** [comment]
Just FYI: I am having some PHP extension issues with 8.3, see also https://github.com/yiisoft/yii2-docker/actions/runs/6183881486/job/16786492326#step:9:4677

**Reply by terabytesoftw** [reply]
The problem seems to be Xdebug not PHP 8.3.

**Reply by schmunk42** [reply]
I re-ran the builds today, see ie. https://github.com/yiisoft/yii2-docker/actions/runs/6183881486/job/17423107578

Images seem to be building now, but there are issues with `HTMLPurifier` and 

``` 
1) yiiunit\framework\data\BaseDataProviderTest::testGenerateId
Calling ReflectionProperty::setValue() with a single argument is deprecated
```

**Reply by terabytesoftw** [reply]
`HtmlPurifier` is compatible with `PHP 8.3` in the `master branch`, we will have to wait for the release, the other error is resolved with `$this->setInnacesibleProperty()`

**Reply by terabytesoftw** [reply]
@schmunk42 You can try again to build the PHP 8.3 image, since htmlpurifier released support for 8.3.

**Reply by schmunk42** [reply]
Imagick is still not available for 8.3 :( https://github.com/yiisoft/yii2-docker/actions/runs/6957432243/job/18930319348
See also https://github.com/mlocati/docker-php-extension-installer#supported-php-extensions

Strangely, I can build the 8.3 image on my machine - dunno what's the issue here.

**Reply by terabytesoftw** [reply]
https://github.com/Imagick/imagick/pull/616 comming soon

**Reply by schmunk42** [reply]
Images build now, see https://github.com/yiisoft/yii2-docker/actions/runs/6971690211/job/18972283148

There's only one failed test:

```
1) yiiunit\framework\data\BaseDataProviderTest::testGenerateId
Calling ReflectionProperty::setValue() with a single argument is deprecated
```

**Reply by terabytesoftw** [reply]
https://github.com/yiisoft/yii2/pull/20074 done :)

**Reply by terabytesoftw** [reply]
@schmunk42 The test is already fixed, now the image can generate.

**Reply by schmunk42** [reply]
Here you go :)
https://hub.docker.com/r/yiisoftware/yii2-php/tags?page=1&name=8.3

I've also triggered a full build, but it'll need about 1.5h

**Reply by terabytesoftw** [reply]
Awesome, thks,

**Reply by schmunk42** [reply]
Doesn't build on `arm64` :confused:  - https://github.com/yiisoft/yii2-docker/actions/runs/6979511769/job/18992949588#step:13:5408

I'll drop a note at https://github.com/Imagick/imagick/pull/616

**Comment by bizley** [comment]
It looks like basically we will do that same thing that @rob006 is suggesting but we will brand it differently - 2.2 will be our 2.0.50-ish with PHP version bumped and all versions after 2.0.49-ish will get only bug and security fixes.

**Reply by rob006** [reply]
Are you sure? I've seen a disturbing amount of "no worry about BC breaks" messages regarding 2.2 version. And if we don't plan to break BC, then releasing it as 2.2.0 will only create compatibility problems with extensions that uses `~2.0.0` as constraint for Yii dependency.

**Reply by bizley** [reply]
Hm, yes, you are right as always but I think this way is better since we will make the bump a conscious decision. With 2.0.x we'll for sure get attacked with "extension x" does not work only because their php constraint is too open.

**Reply by schmunk42** [reply]
We're somewhere stuck in-between, I think every possible BC break should be evaluated and the decision should be made carefully.
Some breaks might be "fixable" with a workaround or some kind of compatibility layer or polyfills.

But in general BC breaks in 2.2 should be as few as possible and not just for some code-candy.

**Reply by rob006** [reply]
> because their php constraint is too open

Yii 2.0.x does not have upper limit for PHP (uses `>=5.4.0` as a constraint), so it never protected you from too wide PHP constraints in extension. Releasing it as 2.2.x will only create 2 additional problems:
1. People with `~2.0.0` as a constraint won't realize that the need to change it to `~2.2.0` and they will stuck on Yii 2.0.49.
2. People will stuck in limbo when they need PHP 9 compatibility, but they can't use Yii 2.2.x because they rely on extensions with `~2.0.0` as constraint for Yii.

**Reply by schmunk42** [reply]
If you know what you are doing you can use `2.2.0 as 2.0.999` as a constraint.

There's no perfect solution regarding the constraints, but having a `2.2` with some degrees of freedom is IMHO the best trade-off.

[addon]
> Yii 2.0.x does not have upper limit for PHP (uses >=5.4.0 as a constraint), so it never protected you from too wide PHP constraints in extension.

Unbound constraints are always a problem, but this applies to so many packages that you can't really maintain it automatically. And, most importantly, you can't fix it for older versions which are already published, besides completely removing them.

**Reply by rob006** [reply]
> If you know what you are doing you can use 2.2.0 as 2.0.999 as a constraint.

That will create problems if some extensions will update their constraint to `~2.2.0` :D

My point is, if we only want to bump required PHP version without actual BC breaks, then it is not worth to release it as 2.2.0. I'm sure 2.2.0 will divide Yii 2 ecosystem, so it would be better to avoid it unless there are some clear benefits of it.

**Reply by bizley** [reply]
If 2.2 goes as some port between v2 and v3 I would keep it as 2.2 (and I would bump php to 7.3 (7.4?) on the 2.0). If the "port" is not a thing we could indeed only do the bump on 2.0.

**Reply by schmunk42** [reply]
> I'm sure 2.2.0 will divide Yii 2 ecosystem

To be honest, the Yii2 extension ecosystem is not very active anymore - sadly. 
You could also argue, that this could help breathe life into the ecosystem. But your points are valid.

Personally or for our company, we'll try to move to Yii 3 - sometime :) We basically have to fork 50% of 3rd party extensions anyway now.

**Reply by terabytesoftw** [reply]
I suggest making a list of the extensions to fork, and we can move it to yiicontrib for example where we can help maintain them among several.

**Reply by mtangoo** [reply]
https://github.com/hosannahighertech/yii2-dynamicform

**Comment by mtangoo** [comment]
On this, AFAIU, 2.2 is supposed to be a bridge to 3.x as well as adding new BC compatibilities. 
To be able to do that we should free 2.2 from 2.0.x

**Comment by terabytesoftw** [comment]
If you agree, i will make a draft of uploading the master branch as proposed by @rob006 almost `PHP 7.3`, we must eliminate the `PHPUNIT` patch.

**Comment by terabytesoftw** [comment]
@cebe @samdark @machour @rhertogh @mtangoo @SilverFire @schmunk42 @bizley @rob006 

1. 7.3
2. 7.4

https://github.com/yiisoft/yii2/pull/19959

What do you think about uploading the main version of the framework, in the master branch to `7.3` or `7.4` at least, will that allow us to remove the patches, add support for PHP `8.3` and continue moving forward?

**Reply by bizley** [reply]
all for 7.4

**Reply by rob006** [reply]
What would be a benefit of using PHP 7.4 instead of PHP 7.3?

**Reply by terabytesoftw** [reply]
1. Perfomance.
2. Type propierties.
3. More support for extension in PHP.
4. In addition to being the latest version 7.
5. New install.

![image](https://github.com/yiisoft/yii2/assets/42547589/035151f6-7ea3-41cb-83cb-ec80dece9cd4)

**Reply by rob006** [reply]
@terabytesoftw You need to be more specific. 

1. Required PHP version does not affect performance - it does not matter what is minimum required PHP version by framework, you can still run it on PHP 7.4 or PHP 8.
2. We can't make exiting properties typed - it is a BC break.
3. I have no idea what do you mean.
4. This sounds really irrelevant.
5. I'm not sure what do you mean by this, but on https://packagist.org/packages/yiisoft/yii2/php-stats PHP 7.3 has over 5% of installs. PHP 7.2 and lower have about 10% of installs. So dropping PHP 7.3 would would cut 50% more people from bugfixes released only in 2.0.x line.

**Reply by terabytesoftw** [reply]
In my opinion, PHP 7.4 can offer significant benefits, even when adding type hints to the code without breaking backward compatibility. Typing of the code is supposed to be done via PHPDoc, and although I know from experience that it's not 100% accurate, it represents a significant advancement. Furthermore, many PHP extensions still support 7.4 more than 7.3. If you're starting a new project, it's ideal to use versions closer to 8.0, such as 7.4.

Of course it is not a silver bullet, but at least it is 7.3 or 7.4 and being able to eliminate the patches for phpunit is already a great advance.

Note: You can see that yii2-queue is difficult to keep tests working correctly with very old versions of PHP.

**Reply by rob006** [reply]
Yii extensions can (and should) define their PHP requirements independently. `yii2-queue` could set their PHP requirement to PHP 8.1 if that would be required, there is no need to force it through framework.

**Reply by terabytesoftw** [reply]
@rob006 The applications have php `7.4`, some extensions `7.4`, would it make sense for all the extensions to have `7.4` ?

**Reply by rob006** [reply]
This is unrelated IMO. App templates are dedicated for new projects, so it make sense to target never version of PHP. But framework and extensions are mostly used by existing apps, so the lower the required PHP is, the more people will be able to use it.

**Reply by terabytesoftw** [reply]
Well, I don't really understand why I should update my legacy application. Shouldn't I keep all my dependencies the same? If I want to start a new project, should I use the latest PHP versions? Currently, many extensions and tests fail from PHP 8.0 and above. It will also be challenging to test older PHP versions, as CI systems work with newer controllers and OS versions. This will make maintenance and bug fixing more difficult. At some point, the framework will lack tests and slowly fade away. Those are my two cents.


For example, the framework supports 8.1 and 8.2, but due to the lack of updated tests in the extensions, some inconvenient issues may arise, as seen in yii2-authclient:

```
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 375
PHP Deprecated:  strpos(): Passing null to parameter #1 ($haystack) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 338
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 350
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 375
PHP Deprecated:  strpos(): Passing null to parameter #1 ($haystack) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 338
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 350
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 375
PHP Deprecated:  strpos(): Passing null to parameter #1 ($haystack) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 338
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 350
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 375
PHP Deprecated:  strpos(): Passing null to parameter #1 ($haystack) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 338
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 350
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 375
PHP Deprecated:  strpos(): Passing null to parameter #1 ($haystack) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 338
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 350
PHP Deprecated:  rawurlencode(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/src/OAuth1.php on line 375
PHP Deprecated:  trim(): Passing null to parameter #1 ($string) of type string is deprecated in /home/runner/work/yii2-authclient/yii2-authclient/vendor/yiisoft/yii2/web/UrlManager.php on line 4[16](https://github.com/yiisoft/yii2-authclient/actions/runs/6357266711/job/17268159083?pr=379#step:9:17)
```

And as you can see, deprecations are not only in the extension but also in the framework itself.

**Reply by forwzb** [reply]
I really hope to do this in yi 3, make yi 3, and then make some changes to 2, because there are too many changes in 3, even if 2 is compatible, the final iteration will update the project code, Ythis. I think yi 3 should be sent first.

**Comment by terabytesoftw** [comment]
https://github.com/yiisoft/yii2-queue/issues/473#issuecomment-1731467446 @s1lver

**Reply by s1lver** [reply]
I can add more to this comment)

Not time long ago (in May) I restored the launch of tests in the yii2-queue package. I encountered problems with the loss of support for Debian and Composer images, and not to mention the fact that I have to sit on old driver packages when they are far ahead. I must say it was not pleasant)

For this package, I would suggest locking the version, with all the latest fixes, to 2.3.6. The 2.3.* branch will remain at the maximum version of PHP 8.0. And starting with version 2.4.0, make the minimum version of PHP 8.1.

**Reply by terabytesoftw** [reply]
`7.4` would not serve as the minimum PHP version ?

**Reply by s1lver** [reply]
Some dependent packages will not allow this in order to have the latest fixes. Well, here, I think that there is no need to apply a general restriction for PHP 7.4

**Reply by s1lver** [reply]
You can, of course, make the minimum version 8.0, but it’s still better to leave 8.0 as the final version for the 2.3.* branch
